### PR TITLE
[FIRRTL] Fix Twine usage in `FIRParseAsserts`

### PR DESCRIPTION
**Pull request by youngar**
When parsing asserts we have a helper class, `ExtractionSummaryCursor`,
to keep track of error messages.  The class holds a `const Twine &`
which created an issue with dangling references.  The following function
was directly creating a dangling reference by passing in `{}` for the
second argument.

```c++
template <typename JsonType>
ExtractionSummaryCursor<JsonType>
makeExtractionSummaryCursor(Location loc, JsonType jsonValue) {
  return ExtractionSummaryCursor<JsonType>{loc, {}, jsonValue};
}
} // namespace
```

By changing the member type to a regular `Twine`, the class is
effectively a wrapper around `Twine` and has the same usage
restrictions. This fixes a bug originally reported by Valgrind.

### CODE REVIEW

**Review by fabianschuiki — APPROVED** [review]
This is an amazing find, thanks a lot @youngar! I keep getting `Twine` wrong :disappointed:

**Comment by lattner** [comment]
Twine is a major footgun!

Is this really enough to solve the problem though?  A twine can (in general) be a binary tree of temporaries, with no strong references to the child nodes.  Will ExtractionSummaryCursor outlive those temporaries?

**Comment by youngar** [comment]
I pushed an update which made the `ExtractionSummaryCursor` constructor and other functions private, preventing people from creating an instance or assigning them to temporaries. I changed `makeExtractionSummaryCursor` stop returning a `ExtractionSummaryCursor` anymore, and instead takes a lambda similar to the `withArray` kinds of functions. All this stuff is to limit who can create a `ExtractionSummaryCursor` to only a handful of functions that we know will keep the Twine temporary alive.

File: lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp [hunk 1]
```diff
@@ -134,10 +153,9 @@ struct ExtractionSummaryCursor {
...
 
 /// Convenience function to create a `ExtractionSummaryCursor`.
-template <typename JsonType>
-ExtractionSummaryCursor<JsonType>
-makeExtractionSummaryCursor(Location loc, JsonType jsonValue) {
-  return ExtractionSummaryCursor<JsonType>{loc, {}, jsonValue};
+template <typename JsonType, typename FnType>
+ParseResult parseJson(Location loc, const JsonType &jsonValue, FnType fn) {
+  return fn(ExtractionSummaryCursor<JsonType>{loc, {}, jsonValue});
 }
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp:159]: I like this lambda continuation style pattern where we have precise control over the lifetime of the parameters passed into the lambda. Very Rust-y :crab: :grin:

**Review by fabianschuiki — APPROVED** [review]
Looks good to me!

**Pull request merged**